homeassistant/components/decora_wifi/__init__.py

```diff
@@ -1 +1,80 @@
 """The decora_wifi component."""
+
+import logging
+
+from decora_wifi import DecoraWiFiSession
+from decora_wifi.models.person import Person
+import voluptuous as vol
+
+from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
+from homeassistant.const import (
+    CONF_PASSWORD,
+    CONF_USERNAME,
+    EVENT_HOMEASSISTANT_STOP,
+    Platform,
+)
+from homeassistant.core import Event, HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.typing import ConfigType
+
+from .config_flow import BASE_SCHEMA
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+PLATFORMS = [Platform.LIGHT]
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: BASE_SCHEMA},
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up decora_wifi component."""
+
+    if DOMAIN not in config:
+        return True
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config[DOMAIN]
+        )
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
+
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    session = DecoraWiFiSession()
+    user = await hass.async_add_executor_job(lambda: session.login(username, password))
+    if not user:
+        raise ConfigEntryAuthFailed("invalid authentication")
+
+    hass.data[DOMAIN][entry.entry_id] = session
+
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    )
+
+    def logout(event: Event) -> None:
+        try:
+            if session is not None:
+                Person.logout(session)
+        except ValueError:
+            _LOGGER.error("Failed to log out of myLeviton Service")
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, logout)
```

Comment:
With a config entry, it's possible that someone reloads the integration. Should we logout then as well? If so, please check `entry.async_on_unload(..)`

Reply:
Well more like, when do you want to logout? I would imagine when you reload the integration (=logging in again) you also want to logout. So I would logout when the entry unloads. `entry.async_on_unload(logout)`.

I honestly haven't found out when to use `@callback` and when not lmao

Reply:
To clarify the goal here was just to logout when hass shuts down, not on reload.
But I guess in hindsight I would want both.

this one makes sense to me (logout when hass stops).
`hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, logout)`

As for the other,
`entry.async_on_unload(logout)` does as exists does not work I think `async_on_unload` expects speficic coroutine 
```
Argument of type "(event: Event) -> None" cannot be assigned to parameter "func" of type "() -> (Coroutine[Any, Any, None] | None)" in function "async_on_unload"
  Type "(event: Event) -> None" cannot be assigned to type "() -> (Coroutine[Any, Any, None] | 
```


This makes type checker happy but that code does not seem to run o unload
(got it from these docs).
```
async def logout(hass: HomeAssistant, entry: ConfigEntry) -> None:
        try:
            if hass.data[DOMAIN][entry.entry_id] is not None:
                hass.async_add_executor_job(lambda: Person.logout(session))
        except ValueError:
            _LOGGER.error("Failed to log out of myLeviton Service")

    entry.async_on_unload(entry.add_update_listener(logout))
```

I am ok just removing this logout signal as well

Reply:
I don't think it's an issue to not logout when reloading, but honestly I am not an expert in this API either. I am just trying to help with UI config flow migration for this.

Also worth considering what I have here is already the current behavior in the existing integration - 
https://github.com/home-assistant/core/blob/dev/homeassistant/components/decora_wifi/light.py#L90

If we scope this PR to a config flow migration only, 
it seems acceptable to leave this unchanged

Reply:
I do think this is a part of the config flow refactor, as before there was no way to reload, so this change also changes the lifetime from restart to restart to restart to reload. Let me check on how to improve on this with other integrations

Reply:
That's fair. Happy to add - would appreciated if you can help provide a code reference for something similar

Reply:
I'm currently on mobile, if I don't respond with one soon, please remind me